### ISSUE TITLE
Omega doesn't trigger quorum_down script execution

### ISSUE THREAD

**Issue by PhantomPhreak** [acassen/keepalived#1774]
**Describe the bug**

Keepalived doesn't trigger `quorum_down` script execution, when `omega` is enabled. It's happening from time to time, not always.

**To Reproduce**
I have following keepalived config:
```
virtual_server 10.196.111.19 8000 {
  delay_loop 2
  lb_algo wlc
  lb_kind DR
  protocol TCP
  alpha
  omega
  quorum_up /tmp/up.sh
  quorum_down /tmp/down.sh

  real_server 127.0.0.1 8000 {
    weight 1
    HTTP_GET {
      url {
        path /
        status_code 200
      }
      connect_timeout 1
      nb_get_retry 3
      delay_before_retry 1
    }
  }

  real_server 127.0.0.2 8000 {
    weight 1
    HTTP_GET {
      url {
        path /
        status_code 200
      }
      connect_timeout 1
      nb_get_retry 3
      delay_before_retry 1
    }
  }
}
```
When stopping keepalived daemon, `quorum_down` script are not executed from time to time, so far i haven't found any reason of that.

Log example:
```
Nov 11 15:55:22 hostname Keepalived[12138]: Starting Keepalived v2.0.10 (11/12,2018)                                                                                                                        
Nov 11 15:55:22 hostname Keepalived[12138]: Running on Linux 4.19.0-8-amd64 #1 SMP Debian 4.19.98-1 (2020-01-26) (built for Linux 4.18.20)                                                                  
Nov 11 15:55:22 hostname Keepalived[12138]: Command line: '/usr/sbin/keepalived' '--dont-fork'                                                                                                              
Nov 11 15:55:22 hostname Keepalived[12138]: Opening file '/etc/keepalived/keepalived.conf'.                                                                                                                 
Nov 11 15:55:22 hostname Keepalived[12138]: Starting Healthcheck child process, pid=12148                                                                                                                   
Nov 11 15:55:22 hostname Keepalived_healthcheckers[12148]: Opening file '/etc/keepalived/keepalived.conf'.                                                                                                                                                  
Nov 11 15:55:22 hostname Keepalived_healthcheckers[12148]: Initializing ipvs                                                                                                                                
Nov 11 15:55:22 hostname Keepalived_healthcheckers[12148]: Activating healthchecker for service [127.0.0.1]:tcp:8000 for VS [10.196.111.19]:tcp:8000                                                        
Nov 11 15:55:22 hostname Keepalived_healthcheckers[12148]: Activating healthchecker for service [127.0.0.2]:tcp:8000 for VS [10.196.111.19]:tcp:8000                                                        
Nov 11 15:55:22 hostname Keepalived_healthcheckers[12148]: Activating BFD healthchecker
Nov 11 15:55:24 hostname Keepalived_healthcheckers[12148]: HTTP status code success to [127.0.0.2]:tcp:8000 url(1).                                                                                         
Nov 11 15:55:24 hostname Keepalived_healthcheckers[12148]: Remote Web server [127.0.0.2]:tcp:8000 succeed on service.                                                                                       
Nov 11 15:55:24 hostname Keepalived_healthcheckers[12148]: Adding service [127.0.0.2]:tcp:8000 to VS [10.196.111.19]:tcp:8000                                                                               
Nov 11 15:55:24 hostname Keepalived_healthcheckers[12148]: Gained quorum 1+0=1 <= 1 for VS [10.196.111.19]:tcp:8000                                                                                         

==> /tmp/status <==
Wed 11 Nov 2020 03:55:24 PM +07 UP

==> /var/log/syslog <==
Nov 11 15:55:24 hostname Keepalived_healthcheckers[12148]: HTTP status code success to [127.0.0.1]:tcp:8000 url(1).
Nov 11 15:55:24 hostname Keepalived_healthcheckers[12148]: Remote Web server [127.0.0.1]:tcp:8000 succeed on service.
Nov 11 15:55:24 hostname Keepalived_healthcheckers[12148]: Adding service [127.0.0.1]:tcp:8000 to VS [10.196.111.19]:tcp:8000
Nov 11 15:55:26 hostname Keepalived_healthcheckers[12148]: Shutting down service [127.0.0.1]:tcp:8000 from VS [10.196.111.19]:tcp:8000
Nov 11 15:55:26 hostname Keepalived[12138]: Stopping
Nov 11 15:55:26 hostname Keepalived_healthcheckers[12148]: Shutting down service [127.0.0.2]:tcp:8000 from VS [10.196.111.19]:tcp:8000
Nov 11 15:55:26 hostname systemd[1]: Stopping Keepalive Daemon (LVS and VRRP)...
Nov 11 15:55:26 hostname Keepalived_healthcheckers[12148]: Stopped
Nov 11 15:55:26 hostname Keepalived[12138]: Stopped Keepalived v2.0.10 (11/12,2018)
Nov 11 15:55:26 hostname systemd[1]: keepalived.service: Succeeded.
Nov 11 15:55:26 hostname systemd[1]: Stopped Keepalive Daemon (LVS and VRRP).

```

**Expected behavior**
`quorum_down` script are executed **always**, when shutting down keepalived service.
Log examples:
keepalived v1.3.2 (works as expected, no problems observed)
```
Nov 11 16:04:11 hostname systemd[1]: Starting Keepalive Daemon (LVS and VRRP)...
Nov 11 16:04:11 hostname Keepalived[12791]: Starting Keepalived v1.3.2 (11/10,2020)
Nov 11 16:04:11 hostname Keepalived[12791]: Setting default script user to 'keepalived_script', uid:gid 1001:1002                                                                                           
Nov 11 16:04:11 hostname Keepalived[12791]: Opening file '/etc/keepalived/keepalived.conf'.
Nov 11 16:04:11 hostname systemd[1]: Started Keepalive Daemon (LVS and VRRP).
Nov 11 16:04:11 hostname Keepalived[12792]: Starting Healthcheck child process, pid=12793
Nov 11 16:04:11 hostname Keepalived[12792]: Starting VRRP child process, pid=12794
Nov 11 16:04:11 hostname Keepalived_vrrp[12794]: Registering Kernel netlink reflector
Nov 11 16:04:11 hostname Keepalived_vrrp[12794]: Registering Kernel netlink command channel
Nov 11 16:04:11 hostname Keepalived_vrrp[12794]: Registering gratuitous ARP shared channel
Nov 11 16:04:11 hostname Keepalived_vrrp[12794]: Opening file '/etc/keepalived/keepalived.conf'.
Nov 11 16:04:11 hostname Keepalived_vrrp[12794]: Unable to load ipset library - libipset.so.3: cannot open shared object file: No such file or directory
Nov 11 16:04:11 hostname Keepalived_vrrp[12794]: Using LinkWatch kernel netlink reflector...
Nov 11 16:04:11 hostname Keepalived_healthcheckers[12793]: Initializing ipvs
Nov 11 16:04:11 hostname Keepalived_healthcheckers[12793]: Registering Kernel netlink reflector
Nov 11 16:04:11 hostname Keepalived_healthcheckers[12793]: Registering Kernel netlink command channel
Nov 11 16:04:11 hostname Keepalived_healthcheckers[12793]: Opening file '/etc/keepalived/keepalived.conf'.
Nov 11 16:04:11 hostname Keepalived_healthcheckers[12793]: Using LinkWatch kernel netlink reflector...
Nov 11 16:04:11 hostname Keepalived_healthcheckers[12793]: Activating healthchecker for service [127.0.0.1]:8000
Nov 11 16:04:11 hostname Keepalived_healthcheckers[12793]: Activating healthchecker for service [127.0.0.2]:8000
Nov 11 16:04:13 hostname Keepalived_healthcheckers[12793]: HTTP status code success to [127.0.0.2]:8000 url(1).
Nov 11 16:04:13 hostname Keepalived_healthcheckers[12793]: Remote Web server [127.0.0.2]:8000 succeed on service.
Nov 11 16:04:13 hostname Keepalived_healthcheckers[12793]: Adding service [127.0.0.2]:8000 to VS [10.196.111.19]:8000
Nov 11 16:04:13 hostname Keepalived_healthcheckers[12793]: Gained quorum 1+0=1 <= 1 for VS [10.196.111.19]:8000
Nov 11 16:04:13 hostname Keepalived_healthcheckers[12793]: Executing [/tmp/up.sh] for VS [10.196.111.19]:8000

==> /tmp/status <==
Wed 11 Nov 2020 04:04:13 PM +07 UP

==> /var/log/syslog <==
Nov 11 16:04:14 hostname Keepalived_healthcheckers[12793]: HTTP status code success to [127.0.0.1]:8000 url(1).
Nov 11 16:04:14 hostname Keepalived_healthcheckers[12793]: Remote Web server [127.0.0.1]:8000 succeed on service.
Nov 11 16:04:14 hostname Keepalived_healthcheckers[12793]: Adding service [127.0.0.1]:8000 to VS [10.196.111.19]:8000
Nov 11 16:04:16 hostname Keepalived[12792]: Stopping
Nov 11 16:04:16 hostname Keepalived_healthcheckers[12793]: Removing service [127.0.0.1]:8000 from VS [10.196.111.19]:8000
Nov 11 16:04:16 hostname Keepalived_healthcheckers[12793]: Removing service [127.0.0.2]:8000 from VS [10.196.111.19]:8000
Nov 11 16:04:16 hostname Keepalived_healthcheckers[12793]: Executing [/tmp/down.sh] for VS [10.196.111.19]:8000
Nov 11 16:04:16 hostname Keepalived_healthcheckers[12793]: Stopped
Nov 11 16:04:16 hostname systemd[1]: Stopping Keepalive Daemon (LVS and VRRP)...

==> /tmp/status <==
Wed 11 Nov 2020 04:04:16 PM +07 down

==> /var/log/syslog <==
Nov 11 16:04:17 hostname Keepalived_vrrp[12794]: Stopped
Nov 11 16:04:17 hostname Keepalived[12792]: Stopped Keepalived v1.3.2 (11/10,2020)
Nov 11 16:04:17 hostname systemd[1]: keepalived.service: Succeeded.
Nov 11 16:04:17 hostname systemd[1]: Stopped Keepalive Daemon (LVS and VRRP).
```
keepalived v2.0.10
```
Nov 11 15:56:42 hostname systemd[1]: Started Keepalive Daemon (LVS and VRRP).
Nov 11 15:56:42 hostname Keepalived[12210]: Starting Keepalived v2.0.10 (11/12,2018)
Nov 11 15:56:42 hostname Keepalived[12210]: Running on Linux 4.19.0-8-amd64 #1 SMP Debian 4.19.98-1 (2020-01-26) (built for Linux 4.18.20)                                                                  
Nov 11 15:56:42 hostname Keepalived[12210]: Command line: '/usr/sbin/keepalived' '--dont-fork'
Nov 11 15:56:42 hostname Keepalived[12210]: Opening file '/etc/keepalived/keepalived.conf'.
Nov 11 15:56:42 hostname Keepalived[12210]: Starting Healthcheck child process, pid=12215
Nov 11 15:56:42 hostname Keepalived_healthcheckers[12215]: Opening file '/etc/keepalived/keepalived.conf'.                                                                                                  
Nov 11 15:56:42 hostname Keepalived_healthcheckers[12215]: Initializing ipvs
Nov 11 15:56:42 hostname Keepalived_healthcheckers[12215]: Activating healthchecker for service [127.0.0.1]:tcp:8000 for VS [10.196.111.19]:tcp:8000                                                        
Nov 11 15:56:42 hostname Keepalived_healthcheckers[12215]: Activating healthchecker for service [127.0.0.2]:tcp:8000 for VS [10.196.111.19]:tcp:8000                                                        
Nov 11 15:56:42 hostname Keepalived_healthcheckers[12215]: Activating BFD healthchecker
Nov 11 15:56:44 hostname Keepalived_healthcheckers[12215]: HTTP status code success to [127.0.0.2]:tcp:8000 url(1).                                                                                         
Nov 11 15:56:44 hostname Keepalived_healthcheckers[12215]: Remote Web server [127.0.0.2]:tcp:8000 succeed on service.                                                                                       
Nov 11 15:56:44 hostname Keepalived_healthcheckers[12215]: Adding service [127.0.0.2]:tcp:8000 to VS [10.196.111.19]:tcp:8000                                                                               
Nov 11 15:56:44 hostname Keepalived_healthcheckers[12215]: Gained quorum 1+0=1 <= 1 for VS [10.196.111.19]:tcp:8000                                                                                         

==> /tmp/status <==
Wed 11 Nov 2020 03:56:44 PM +07 UP

==> /var/log/syslog <==
Nov 11 15:56:45 hostname Keepalived_healthcheckers[12215]: HTTP status code success to [127.0.0.1]:tcp:8000 url(1).                                                                                         
Nov 11 15:56:45 hostname Keepalived_healthcheckers[12215]: Remote Web server [127.0.0.1]:tcp:8000 succeed on service.                                                                                       
Nov 11 15:56:45 hostname Keepalived_healthcheckers[12215]: Adding service [127.0.0.1]:tcp:8000 to VS [10.196.111.19]:tcp:8000                                                                               
Nov 11 15:56:45 hostname systemd[1]: Stopping Keepalive Daemon (LVS and VRRP)...
Nov 11 15:56:45 hostname Keepalived[12210]: Stopping
Nov 11 15:56:45 hostname Keepalived_healthcheckers[12215]: Shutting down service [127.0.0.1]:tcp:8000 from VS [10.196.111.19]:tcp:8000                                                                      
Nov 11 15:56:45 hostname Keepalived_healthcheckers[12215]: Shutting down service [127.0.0.2]:tcp:8000 from VS [10.196.111.19]:tcp:8000                                                                      

==> /tmp/status <==
Wed 11 Nov 2020 03:56:45 PM +07 down

==> /var/log/syslog <==
Nov 11 15:56:45 hostname Keepalived_healthcheckers[12215]: Stopped
Nov 11 15:56:45 hostname Keepalived[12210]: Stopped Keepalived v2.0.10 (11/12,2018)
Nov 11 15:56:45 hostname systemd[1]: keepalived.service: Succeeded.
Nov 11 15:56:45 hostname systemd[1]: Stopped Keepalive Daemon (LVS and VRRP).
```

**Keepalived version**
v2.0.10 (upstream version for Debian 10)
v2.0.20
v2.1.5

```
# keepalived -v
Keepalived v2.0.10 (11/12,2018)

Copyright(C) 2001-2018 Alexandre Cassen, <acassen@gmail.com>

Built with kernel headers for Linux 4.18.20
Running on Linux 4.19.0-8-amd64 #1 SMP Debian 4.19.98-1 (2020-01-26)

configure options: --build=x86_64-linux-gnu --prefix=/usr --includedir=${prefix}/include --mandir=${prefix}/share/man --infodir=${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --disable-silent-rules --libdir=${prefix}/lib/x86_64-linux-gnu --libexecdir=${prefix}/lib/x86_64-linux-gnu --runstatedir=/run --disable-maintainer-mode --disable-dependency-tracking --with-kernel-dir=debian/ --enable-snmp --enable-sha1 --enable-snmp-rfcv2 --enable-snmp-rfcv3 --enable-dbus --enable-dbus-create-instance --enable-json --enable-bfd build_alias=x86_64-linux-gnu CFLAGS=-g -O2 -fdebug-prefix-map=/build/keepalived-8agDac/keepalived-2.0.10=. -fstack-protector-strong -Wformat -Werror=format-security LDFLAGS=-Wl,-z,relro CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2

Config options:  LIBIPSET_DYNAMIC LVS VRRP VRRP_AUTH JSON BFD OLD_CHKSUM_COMPAT FIB_ROUTING SNMP_V3_FOR_V2 SNMP_VRRP SNMP_CHECKER SNMP_RFCV2 SNMP_RFCV3 DBUS DBUS_CREATE_INSTANCE

System options:  PIPE2 SIGNALFD INOTIFY_INIT1 VSYSLOG EPOLL_CREATE1 IPV4_DEVCONF LIBNL3 RTA_ENCAP RTA_EXPIRES RTA_NEWDST RTA_PREF FRA_SUPPRESS_PREFIXLEN FRA_SUPPRESS_IFGROUP FRA_TUN_ID RTAX_CC_ALGO RTAX_QUICKACK RTEXT_FILTER_SKIP_STATS FRA_L3MDEV FRA_UID_RANGE RTAX_FASTOPEN_NO_COOKIE RTA_VIA FRA_OIFNAME FRA_PROTOCOL FRA_IP_PROTO FRA_SPORT_RANGE FRA_DPORT_RANGE RTA_TTL_PROPAGATE IFA_FLAGS IP_MULTICAST_ALL LWTUNNEL_ENCAP_MPLS LWTUNNEL_ENCAP_ILA LIBIPTC LIBIPSET_PRE_V7 LIBIPVS_NETLINK IPVS_DEST_ATTR_ADDR_FAMILY IPVS_SYNCD_ATTRIBUTES IPVS_64BIT_STATS VRRP_VMAC SOCK_NONBLOCK SOCK_CLOEXEC O_PATH GLOB_BRACE INET6_ADDR_GEN_MODE VRF SO_MARK SCHED_RT SCHED_RESET_ON_FORK

```

**Distro (please complete the following information):**
 - Debian
 - 10.3
 - x86_64

**Notify and track scripts**
```
$ cat /tmp/up.sh
#!/bin/bash
/usr/bin/echo "`/usr/bin/date` UP" >> /tmp/status
$ cat /tmp/down.sh 
#!/bin/bash
/usr/bin/echo "`/usr/bin/date` down" >> /tmp/status
```

**Comment by pqarmitage** [comment]
Currently when the shutdown of keepalived starts it runs the notify script for each VS with omega set. It then waits 1 second, and then terminates any scripts that haven't completed. It doesn't look, however, as though you are experiencing a problem with the wait of 1 second being too short.

I can see that the log for v1.3.2 has `Executing [/tmp/down.sh] for VS [10.196.111.19]:8000` whereas the log for v2.0.10 does not. That however is not a problem since keepalived no longer logs when it runs the script.

What I can see from the test is that for both v1.3.2 and v2.0.10 the /tmp/status file contains `down` at the end, so the information above isn't showing the problem, and there isn't anything showing UP after keepaived is shutdown.

**Comment by pqarmitage** [comment]
I don't think this relates to your problem, but I found that if `lvs_flush_on_stop` is configured keepalived didn't send notifies on shutdown. Commit c9b6d02 resolves this.

It might be worth adding the following patch, so the keepalived logs when sending notifies when stopping, so that you can see if it is missing sending them:
```
diff --git a/keepalived/check/ipwrapper.c b/keepalived/check/ipwrapper.c
index 80270a1d..08641808 100644
--- a/keepalived/check/ipwrapper.c
+++ b/keepalived/check/ipwrapper.c
@@ -234,6 +234,9 @@ update_vs_notifies(virtual_server_t *vs, bool stopping)
             (!weight_sum || weight_sum < threshold))) {
                vs->quorum_state_up = false;
                do_vs_notifies(vs, false, threshold, weight_sum, stopping);
+
+if (stopping)
+log_message(LOG_INFO, "(%s) - sent vs down notify", FMT_VS(vs));
        }
 }
 
```

Since notify_quorum_up and and notify_quorum_down are scripts, then they are not entirely reliable, for example if there is a delay in their execution being scheduled. If you can, it would be better to use a notify_fifo.

**Comment by PhantomPhreak** [comment]
>  It doesn't look, however, as though you are experiencing a problem with the wait of 1 second being too short.

Exactly.

> I can see that the log for v1.3.2 has Executing [/tmp/down.sh] for VS [10.196.111.19]:8000 whereas the log for v2.0.10 does not. That however is not a problem since keepalived no longer logs when it runs the script.

Yes, it relates to logging only

> What I can see from the test is that for both v1.3.2 and v2.0.10 the /tmp/status file contains down at the end, so the information above isn't showing the problem, and there isn't anything showing UP after keepaived is shutdown.

As i described, this problem is not reproducible all the time for v2.0.10: sometimes `quorum_down` is triggered when keepalived shutting down, sometimes not. I provided log examples for both cases for v2.0.10 (when problem occurs, and for **expected behavior**)
So, as shown, there are no `down` when problem occurs, means that `quorum_down` script wasn't executed.

**Comment by PhantomPhreak** [comment]
> I don't think this relates to your problem, but I found that if lvs_flush_on_stop is configured keepalived didn't send notifies on shutdown. Commit c9b6d02 resolves this.

Yes, i it's not related to the `lvs_flush_on_stop`, nevertheless, thanks for pointing out this issue, i'll mind it.

Thanks for the idea about patch and `notify_fifo`, gonna try it.

**Comment by pqarmitage** [comment]
@PhantomPhreak Have you been able to make any progress with this?

**Comment by PhantomPhreak** [comment]
@pqarmitage not yet, even if `notify_fifo` works, it can be considered as a worksome, since original problem still exists.

**Comment by pqarmitage** [comment]
The reason I added the notify fifos was due to the fact that using scripts is unreliable, particularly regarding the order of execution. Using notify fifos ensures that messages are delivered in the correct order, whereas when executing scripts, the sequence of execution is random. Of course, that doesn't apply here, but the same principle applies. Scripts are unreliable whereas fifos are reliable.

I don't think using notify_fifos should be considered as a workaround; they are a solution to the problem of running scripts being non-deterministic. For example if real server 127.0.0.1is up and RS 127.0.0.2 is down, and then the HTTP_GET fails for 127.0.0.1 and then immediately succeeds for 127.0.0.2, when the HTTP_GET for 127.0.0.1 fails the `/tmp/down.sh` script is forked, and then when the HTTP_GET for 127.0.0.2 succeeds the `/tmp/up.sh` script is forked. These scripts now run independently and concurrently and either of them could write to `/tmp/status` first (or more to the point either could run second).

I think I can see now what is probably happening here. Keepalived uses the prctl() system call to set the signal that any child process of keepalived should receive when keepalived terminates (this is to ensure that random processes aren't left lying around when keepalived terminates). If keepalived terminates before the script has completed, the script will be sent a SIGTERM signal, which will by default cause the script to terminate. I suspect that what is happening here is that sometimes the script does not complete executing until keepalived terminates, and therefore the script terminates when it receives a SIGTERM. As explained above, the order of execution (of keepalived and the scripts) is non-deterministic, and hence the use of scripts is unreliable.

If you want to continue using scripts, the first thing to try is to add 
`trap "" TERM`
as the second line of your down script. I'm not sure that this is a complete solution, because I think bash could be terminated before the script even starts executing.

However, I think using a fifo is a reliable way to do this. In the doc/samples directory of the source tree is a sample fifo reading script, `sample_notify_fifo.sh`. This handles the SIGTERM that is sent when keepalived is exiting and runs for a further 0.5 seconds before it exits. All that you would need to do to use the script is to modify the lines immediately following the `# Now take whatever action is required` comments. If you specify the script using the `lvs_notify_fifo_script` global_defs option, then keepalived will run the script when it starts.

Other possibilities (which require source code modifications) are:
1. Add a delay between keepalived executing the scripts and keepalived terminating. Again this is non-deterministic but it may help.
1. keepalived reduces its priority, calls sched_yield() and uses non realtime scheduling when exiting so that scripts are more likely to run. This won't help on a multi core system that isn't busy.
1. keepalived waits for notify scripts to terminate (I am not inclined to implement this since notify scripts are fire and forget).
1. Implement `notify_file` so that keepalived itself writes a string to a file.

However, since there is a reliable mechanism, by using a fifo and a script to read the fifo, I am not convinced that we need to implement any of the above 4 ideas.
